File: ruby/lib/steam/packets/steam_packet_factory.rb
# This code is free software; you can redistribute it and/or modify it under
# the terms of the new BSD License.
#
# Copyright (c) 2008-2011, Sebastian Staudt

require 'exceptions/steam_condenser_exception'
require 'steam/packets/s2a_info_detailed_packet'
require 'steam/packets/a2s_info_packet'
require 'steam/packets/s2a_info2_packet'
require 'steam/packets/a2s_player_packet'
require 'steam/packets/s2a_player_packet'
require 'steam/packets/a2s_rules_packet'
require 'steam/packets/s2a_rules_packet'
require 'steam/packets/a2s_serverquery_getchallenge_packet'
require 'steam/packets/s2c_challenge_packet'
require 'steam/packets/a2m_get_servers_batch2_packet'
require 'steam/packets/m2a_server_batch_packet'
require 'steam/packets/m2c_isvalidmd5_packet'
require 'steam/packets/m2s_requestrestart_packet'
require 'steam/packets/s2a_logstring_packet'
require 'steam/packets/rcon/rcon_goldsrc_response'

# This module provides functionality to handle raw packet data, including
# data split into several UDP / TCP packets and BZIP2 compressed data. It's the
# main utility to transform data bytes into packet objects.
#
# @author Sebastian Staudt
# @see SteamPacket
module SteamPacketFactory

  # Creates a new packet object based on the header byte of the given raw data
  #
  # @param [String] raw_data The raw data of the packet
  # @raise [SteamCondenserException] if the packet header is not recognized
  # @return [SteamPacket] The packet object generated from the packet data
  def self.packet_from_data(raw_data)
    header = raw_data[0].ord
    data = raw_data[1..-1]

    case header
      when SteamPacket::S2A_INFO_DETAILED_HEADER
        return S2A_INFO_DETAILED_Packet.new(data)
      when SteamPacket::A2S_INFO_HEADER
        return A2S_INFO_Packet.new
      when SteamPacket::S2A_INFO2_HEADER
        return S2A_INFO2_Packet.new(data)
      when SteamPacket::A2S_PLAYER_HEADER
        return A2S_PLAYER_Packet.new
      when SteamPacket::S2A_PLAYER_HEADER
        return S2A_PLAYER_Packet.new(data)
      when SteamPacket::A2S_RULES_HEADER
        return A2S_RULES_Packet
      when SteamPacket::S2A_RULES_HEADER
        return S2A_RULES_Packet.new(data)
      when SteamPacket::A2S_SERVERQUERY_GETCHALLENGE_HEADER
        return A2S_SERVERQUERY_GETCHALLENGE_Packet.new
      when SteamPacket::S2C_CHALLENGE_HEADER
        return S2C_CHALLENGE_Packet.new(data)
      when SteamPacket::A2M_GET_SERVERS_BATCH2_HEADER
        return A2M_GET_SERVERS_BATCH2_Packet.new(data)
      when SteamPacket::M2A_SERVER_BATCH_HEADER
        return M2A_SERVER_BATCH_Packet.new(data)
      when SteamPacket::M2C_ISVALIDMD5_HEADER
        return M2C_ISVALIDMD5_Packet.new(data)
      when SteamPacket::M2S_REQUESTRESTART_HEADER
        return M2S_REQUESTRESTART_Packet.new(data)
      when SteamPacket::RCON_GOLDSRC_CHALLENGE_HEADER,
           SteamPacket::RCON_GOLDSRC_NO_CHALLENGE_HEADER,
           SteamPacket::RCON_GOLDSRC_RESPONSE_HEADER
        return RCONGoldSrcResponse.new(data)
      when SteamPacket::S2A_LOGSTRING_HEADER
        return S2A_LOGSTRING_Packet.new(data)
      else
        raise SteamCondenserException.new("Unknown packet with header 0x#{header.to_s(16)} received.")
    end
  end

  # Reassembles the data of a split and/or compressed packet into a single
  # packet object
  #
  # @param [Array<String>] split_packets An array of packet data
  # @param [true, false] is_compressed whether the data of this packet is
  #        compressed
  # @param [Fixnum] packet_checksum The CRC32 checksum of the decompressed
  #        packet data
  # @raise [SteamCondenserException] if the bz2 gem is not installed
  # @raise [PacketFormatException] if the calculated CRC32 checksum does not
  #        match the expected value
  # @return [SteamPacket] The reassembled packet
  # @see packet_from_data
  def self.reassemble_packet(split_packets, is_compressed = false, packet_checksum = 0)
    packet_data = split_packets.join ''

    if is_compressed
      require 'zlib'

      begin
        packet_data = BZ2.uncompress(packet_data)
      rescue LoadError
        raise SteamCondenserException.new('You need to install the libbzip2 interface for Ruby.')
      end

      unless Zlib.crc32(packet_data) == packet_checksum
        raise PacketFormatException.new('CRC32 checksum mismatch of uncompressed packet data.')
      end
    end

    packet_from_data packet_data[4..-1]
  end

end

File: php/lib/steam/packets/SteamPacketFactory.php
<?php
/**
 * This code is free software; you can redistribute it and/or modify it under
 * the terms of the new BSD License.
 *
 * Copyright (c) 2008-2011, Sebastian Staudt
 *
 * @author     Sebastian Staudt
 * @license    http://www.opensource.org/licenses/bsd-license.php New BSD License
 * @package    Steam Condenser (PHP)
 * @subpackage Packets
 */

require_once STEAM_CONDENSER_PATH . 'exceptions/PacketFormatException.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/rcon/RCONGoldSrcResponse.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/S2A_INFO_DETAILED_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/A2S_INFO_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/S2A_INFO2_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/A2S_PLAYER_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/S2A_PLAYER_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/A2S_RULES_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/S2A_RULES_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/A2S_SERVERQUERY_GETCHALLENGE_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/S2C_CHALLENGE_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/A2M_GET_SERVERS_BATCH2_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/M2A_SERVER_BATCH_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/M2C_ISVALIDMD5_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/M2S_REQUESTRESTART_Packet.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/S2A_LOGSTRING_Packet.php';

/**
 * @package    Steam Condenser (PHP)
 * @subpackage Packets
 */
abstract class SteamPacketFactory
{
	/**
	 * @param byte[] $rawData
	 * @return SteamPacket
	 */
	public static function getPacketFromData($rawData)
	{
		$header = ord($rawData[0]);
		$data = substr($rawData, 1);

		switch($header)
		{
			case SteamPacket::S2A_INFO_DETAILED_HEADER:
				return new S2A_INFO_DETAILED_Packet($data);

			case SteamPacket::A2S_INFO_HEADER:
				return new A2S_INFO_Packet();

			case SteamPacket::S2A_INFO2_HEADER:
				return new S2A_INFO2_Packet($data);

			case SteamPacket::A2S_PLAYER_HEADER:
				return new A2S_PLAYER_Packet();

			case SteamPacket::S2A_PLAYER_HEADER:
				return new S2A_PLAYER_Packet($data);

			case SteamPacket::A2S_RULES_HEADER:
				return new A2S_RULES_Packet();

			case SteamPacket::S2A_RULES_HEADER:
				return new S2A_RULES_Packet($data);

			case SteamPacket::A2S_SERVERQUERY_GETCHALLENGE_HEADER:
				return new A2S_SERVERQUERY_GETCHALLENGE_Packet();

			case SteamPacket::S2C_CHALLENGE_HEADER:
				return new S2C_CHALLENGE_Packet($data);

			case SteamPacket::A2M_GET_SERVERS_BATCH2_HEADER:
				return new A2M_GET_SERVERS_BATCH2_Packet($data);

			case SteamPacket::M2A_SERVER_BATCH_HEADER:
				return new M2A_SERVER_BATCH_Packet($data);

            case SteamPacket::M2C_ISVALIDMD5_HEADER:
                return new M2C_ISVALIDMD5_Packet($data);

            case SteamPacket::M2S_REQUESTRESTART_HEADER:
                return new M2S_REQUESTRESTART_Packet($data);

			case SteamPacket::RCON_GOLDSRC_CHALLENGE_HEADER:
            case SteamPacket::RCON_GOLDSRC_NO_CHALLENGE_HEADER:
			case SteamPacket::RCON_GOLDSRC_RESPONSE_HEADER:
				return new RCONGoldSrcResponse($data);

            case SteamPacket::S2A_LOGSTRING_HEADER:
                return new S2A_LOGSTRING_Packet($data);

			default:
				throw new PacketFormatException("Unknown packet with header 0x" . dechex($header) . " received.");
		}
	}

	public static function reassemblePacket($splitPackets, $isCompressed = false, $packetChecksum = 0)
	{
		$packetData = "";

		foreach($splitPackets as $splitPacket)
		{
			if($splitPacket == null)
			{
				throw new UncompletePacketException();
			}

			$packetData .= $splitPacket;
		}

		if($isCompressed)
		{
			$packetData = bzdecompress($packetData);

			if(crc32($packetData) != $packetChecksum)
			{
				throw new PacketFormatException("CRC32 checksum mismatch of uncompressed packet data.");
			}
		}

		// Omit leading 0xFFFFFFFF
		$packetData = substr($packetData, 4);

		return self::getPacketFromData($packetData);
	}
}
?>
